src/bin/wasm-tools/wast.rs

```diff
@@ -0,0 +1,584 @@
+use wast::parser::{self, ParseBuffer};
+use wast::{QuoteWat, Wast, WastDirective, Wat};
+
+/// A subcommand to test `*.wast` files short of executing WebAssembly code.
+///
+/// This subcommand can be used to validate the `*.wast` test format used by the
+/// WebAssembly spec interpreter. This command takes a single test case as input
+/// and it will validate as much as it can about it including:
+///
+/// * The syntax of the test is valid.
+/// * Modules that are declared to be valid are indeed valid.
+/// * Modules that are declared to be invalid are indeed invalid.
+///
+/// This subcommand does not actually execute any WebAssembly code, so the
+/// `*.wast` test is not guaranteed to pass in a real runtime if passed to a
+/// runtime. This can be used as a quick check to double-check that a file is
+/// almost all valid, however.
+///
+/// The `wasm-tools` project itself also heavily relies on this subcommand for
+/// internal testing, so many tests for `wasm-tools` use this subcommand as a
+/// way to write tests.
+#[derive(Parser)]
+pub struct Opts {
+    #[clap(flatten)]
+    general: wasm_tools::GeneralOpts,
+
+    /// Input file to process.
+    ///
+    /// If not provided or if this is `-` then stdin is read entirely and
+    /// processed.
+    input: Option<PathBuf>,
+
+    /// Whether to ignore the error message expectation in `(assert_invalid ..)`
+    /// directives.
+    ///
+    /// When enabled this will only assert that modules produce an error, not
+    /// that the error produced by wasm-tools matches that written in the test
+    /// itself.
+    #[clap(long)]
+    ignore_error_messages: bool,
+
+    #[clap(flatten)]
+    features: CliFeatures,
+
+    /// Whether or not "confusing unicode characters" are allowed to be present
+    /// in `*.wast` files.
+    ///
+    /// The default for this is `false`.
+    #[clap(long)]
+    allow_confusing_unicode: bool,
+
+    /// Perform extra internal assertions in `wasm-tools` itself over the
+    /// provided test.
+    ///
+    /// This is used within the testing of `wasm-tools` itself to perform more
+    /// tests by default such as asserting that the text format round-trips
+    /// where possible. Asserts can be listed here to perform extra checks when
+    /// validating the `*.wast` input.
+    #[clap(long, short, value_parser = parse_asserts)]
+    assert: Vec<Vec<Assert>>,
+
+    /// Directory to place snapshots in with `--assert snapshot-*` flags.
+    #[clap(long)]
+    snapshot: Option<PathBuf>,
+}
+
+impl Opts {
+    pub fn general_opts(&self) -> &wasm_tools::GeneralOpts {
+        &self.general
+    }
+
+    pub fn run(&self) -> Result<()> {
+        let (test, contents) = self.read_wast()?;
+        let contents = contents.as_str();
+        macro_rules! adjust {
+            ($e:expr) => {{
+                let mut e = wast::Error::from($e);
+                e.set_path(test);
+                e.set_text(contents);
+                e
+            }};
+        }
```

Comment:
This seems like it could just be a closure, and doesn't actually need to be a macro?

Reply:
Oh I guess because it is generic over the type of `$e` it can't be a closure. blech.